File: framework/languages/node/e2e/RuntimeMonitoring/Client/Scenarios/mon-a3-spot-events-scenario.ts
// MON-A3: Channel readiness와 실제 request 결과를 대조한다 시나리오를 검증한다.
import type { EvidenceWaitReq, ProfileReq, ProfileRes } from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { getJson, postJson } from '../../../http-client';
import { waitForRouteStatus, delay } from '../Support/public-status';
import { ensure } from '../Support/scenario-assert';

export async function runMonA3(options: ClientOptions): Promise<void> {
  const ready = await waitForRouteStatus(
    options.serviceUrl,
    (status) => status.channels.some((channel) =>
      channel.channelName === 'monitor.profile' && channel.isReady && channel.readyTargetCount === 1),
    'MON-A3 expected svc-b to be the only ready profile target.'
  );
  const first = await requestProfile(options, 'mon-a3-before-exclude');
  ensure(first.providerRid === 'svc-b' && first.value === 'profile:before-exclude', 'MON-A3 initial request did not reach svc-b.');

  await postJson<object>(options.serviceBUrl, '/admin/exclude', {});
  await waitForWeight(options.serviceBUrl, 0);
  const excluded = await waitForRouteStatus(
    options.serviceUrl,
    (status) => status.channels.some((channel) =>
      channel.channelName === 'monitor.profile' && !channel.isReady && channel.readyTargetCount === 0),
    'MON-A3 Channel status did not remove the excluded target.'
  );
  ensure(BigInt(excluded.sequence) > BigInt(ready.sequence), 'MON-A3 Channel readiness sequence did not advance after weight exclusion.');

  let rejected = false;
  try {
    await requestProfile(options, 'mon-a3-excluded');
  } catch {
    rejected = true;
  }
  ensure(rejected, 'MON-A3 request succeeded while no positive-weight target was ready.');

  await postJson<object>(options.serviceBUrl, '/admin/include', {});
  await waitForWeight(options.serviceBUrl, 100);
  const restored = await waitForRouteStatus(
    options.serviceUrl,
    (status) => status.channels.some((channel) =>
      channel.channelName === 'monitor.profile' && channel.isReady && channel.readyTargetCount === 1),
    'MON-A3 Channel status did not restore the included target.'
  );
  const final = await requestProfile(options, 'mon-a3-after-include');
  ensure(final.providerRid === 'svc-b' && final.value === 'profile:after-include', 'MON-A3 restored request did not reach svc-b.');
  ensure(BigInt(restored.sequence) > BigInt(excluded.sequence), 'MON-A3 Channel sequence did not advance after weight restoration.');

  await postJson<string[]>(options.serviceBUrl, '/evidence/wait', {
    containsAll: ['profile-request|rid=svc-b|marker=mon-a3-after-include|value=after-include'],
    containsAnyGroups: [],
    timeoutMilliseconds: 10000
  } satisfies EvidenceWaitReq);
  console.log('scenario MON-A3 passed');
}

async function requestProfile(options: ClientOptions, marker: string): Promise<ProfileRes> {
  const request: ProfileReq = { value: marker.replace('mon-a3-', ''), marker };
  return await postJson<ProfileRes>(options.triggerUrl, '/profile/request/service-b', request);
}

async function waitForWeight(serviceUrl: string, expected: number): Promise<void> {
  const deadline = Date.now() + 10_000;
  while (Date.now() < deadline) {
    if ((await getJson<{ readonly weight: number }>(serviceUrl, '/admin/weight')).weight === expected) return;
    await delay();
  }
  throw new Error(`MON-A3 service weight did not become ${expected}.`);
}

File: framework/languages/dotnet/e2e/RuntimeMonitoring/Client/Scenarios/MonA3SpotEventsScenario.cs
// Verifies Config 7 MON-A3 ChannelName readiness against actual selection.
using System.Diagnostics;
using RuntimeMonitoring.Client.Support;
using RuntimeMonitoring.Shared;
using Zlink.HttpClient;

namespace RuntimeMonitoring.Client.Scenarios;

internal static class MonA3SpotEventsScenario
{
    public static async Task RunAsync(ClientOptions options)
    {
        using var serviceA = ZLinkHttpClient.Create(options.ServiceUrl)
            .Timeout(TimeSpan.FromSeconds(35)).Build();
        using var serviceB = ZLinkHttpClient.Create(options.ServiceBUrl)
            .Timeout(TimeSpan.FromSeconds(35)).Build();

        AssertChannel(await WaitForReadyTargetsAsync(serviceA, 1), 1);
        AssertChannel(await WaitForReadyTargetsAsync(serviceB, 1), 1);
        var evidenceBaseline =
            (await serviceA.Get("/evidence").Async<string[]>()).Body.Length;
        ZlinkStreamAssert.Ensure(
            await ObserveProviderAsync(serviceA, "svc-b", 8),
            "MON-A3 baseline selection never reached svc-b.");
        ZlinkStreamAssert.Ensure(
            await ObserveProviderAsync(serviceB, "svc-a", 8),
            "MON-A3 remote caller did not select svc-a.");

        await serviceB.Post("/admin/weight/exclude").AsyncRaw();
        AssertChannel(await WaitForReadyTargetsAsync(serviceA, 0), 0);
        AssertChannel(await WaitForReadyTargetsAsync(serviceB, 1), 1);
        await AssertNoSelectableTargetAsync(serviceA);

        var eventEvidence = (await serviceA.Post("/evidence/wait")
            .Body(new EvidenceWaitReq(
                [
                    "identifier=zlink.runtime.mesh_node.channel_changed",
                    $"channel={RuntimeMonitoringNames.Channel}"
                ],
                [],
                TimeoutMilliseconds: 3000,
                AfterIndex: evidenceBaseline))
            .Async<string[]>()).Body;
        ZlinkStreamAssert.Ensure(
            eventEvidence.Any(line =>
                line.Contains("identifier=zlink.runtime.mesh_node.channel_changed",
                    StringComparison.Ordinal)
                && line.Contains(
                    $"channel={RuntimeMonitoringNames.Channel}",
                    StringComparison.Ordinal)),
            "MON-A3 channel_changed event was missing.");

        await serviceB.Post("/admin/weight/include").AsyncRaw();
        AssertChannel(await WaitForReadyTargetsAsync(serviceA, 1), 1);
        AssertChannel(await WaitForReadyTargetsAsync(serviceB, 1), 1);
        ZlinkStreamAssert.Ensure(
            await ObserveProviderAsync(serviceA, "svc-b", 8),
            "MON-A3 restored svc-b did not become selectable.");
        ZlinkStreamAssert.Ensure(
            await ObserveProviderAsync(serviceB, "svc-a", 8),
            "MON-A3 restored remote caller did not select svc-a.");

        Console.WriteLine("scenario MON-A3 passed");
    }

    private static void AssertChannel(
        MeshRuntimeSnapshotRes status,
        int readyTargets)
    {
        var channel = status.Channels.Single(candidate =>
            candidate.ChannelName == RuntimeMonitoringNames.Channel);
        ZlinkStreamAssert.Ensure(
            channel.IsReady == readyTargets > 0
                && channel.ReadyTargetCount == readyTargets,
            $"MON-A3 channel status was ready={channel.ReadyTargetCount},"
            + $" isReady={channel.IsReady}.");
    }

    private static async Task<bool> ObserveProviderAsync(
        ZLinkHttpClient service,
        string expectedRid,
        int attempts)
    {
        for (var attempt = 0; attempt < attempts; attempt++)
        {
            var response = (await service.Post("/profile/request")
                .Body(new ProfileReq($"weight-{attempt}", $"mon-a3-{attempt}"))
                .Async<ProfileRes>()).Body;
            if (response.ProviderRid == expectedRid)
                return true;
        }
        return false;
    }

    private static async Task AssertNoSelectableTargetAsync(
        ZLinkHttpClient service)
    {
        var response = await service.Post("/profile/request")
            .Body(new ProfileReq("weight-0-unavailable", "mon-a3-unavailable"))
            .AsyncRaw();
        ZlinkStreamAssert.Ensure(
            response.Status == 500,
            $"MON-A3 expected a terminal no-target response, got HTTP {response.Status}.");
    }

    private static async Task<MeshRuntimeSnapshotRes> WaitForReadyTargetsAsync(
        ZLinkHttpClient service,
        int expected)
    {
        var elapsed = Stopwatch.StartNew();
        while (true)
        {
            var status = (await service.Get(
                    $"/runtime/snapshot/{RuntimeMonitoringNames.Channel}")
                .Async<MeshRuntimeSnapshotRes>()).Body;
            var channel = status.Channels.Single(candidate =>
                candidate.ChannelName == RuntimeMonitoringNames.Channel);
            if (channel.ReadyTargetCount == expected)
                return status;
            if (elapsed.Elapsed >= TimeSpan.FromSeconds(3))
                throw new InvalidOperationException(
                    "MON-A3 channel status did not reach the expected state.");
            await Task.Delay(TimeSpan.FromMilliseconds(50));
        }
    }
}
